### PR TITLE
add porous solid particles,

### PR DESCRIPTION
**Pull request by Xiaojingtang1234**
add porous solid material
add porous dynamics, momentum constraint dynamics, fluid diffusion dynamics, add test case

### CODE REVIEW

File: tests/user_examples/extra_src/shared/porous_solid_particles.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+/**
+ *	pre-claimed classes
+ */
+//class PorousMediaSolid;
+ 
+/**
+ * @class PorousMediaParticles
+ * @brief A group of particles with elastic body particle data.
+ */
+	class PorousMediaParticles : public ElasticSolidParticles
```
Comment (Xiangyu-Hu) [inline tests/user_examples/extra_src/shared/porous_solid_particles.h:46]: we need to discuss how to add porous media to the code in its simplest but extendable way,

File: tests/user_examples/extra_src/shared/particle_momentum_dissipation.hpp [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            size_t index_j = inner_neighborhood.j_[n - 1];
+            Real mass_j = mass_[index_j];
+
+            VariableType variable_derivative = (variable_i - variable_[index_j]);
+            VariableType increment = parameter_b[n - 1] * variable_derivative / (mass_i * mass_j - parameter_b[n - 1] * (mass_i + mass_j));
+
+            variable_[index_i] += increment * mass_j;
+            variable_[index_j] -= increment * mass_i;
+        }
+	}
```
Comment (DongWuTUM) [inline tests/user_examples/extra_src/shared/particle_momentum_dissipation.hpp:57]: Please, pay attention to the coding format, such as indenting the first letter.

File: tests/user_examples/extra_src/shared/porous_media_dynamics.cpp [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        //three parts for the momentum increment
+        total_momentum_increment += (Stress_[index_i] + Stress_[index_j])* gradw_ijV_j_
+                                    +  numerical_dissipation_factor_  * numerical_stress_ij * weight * gradw_ijV_j_ ;
+    
+        total_momentum_increment -= (outer_fluid_velocity_relative_fluid_flux_[index_i] + outer_fluid_velocity_relative_fluid_flux_[index_j])
+                                     *gradw_ijV_j_;
+    }
+
+    dtotal_momentum_dt_[index_i] = total_momentum_increment;
+}
```
Comment (DongWuTUM) [inline tests/user_examples/extra_src/shared/porous_media_dynamics.cpp:95]: Following the coding style, the interaction should be moved to .h file.

File: tests/user_examples/extra_src/shared/porous_media_dynamics.cpp [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        Matd numerical_stress_ij = 0.5 * (F_[index_i] + F_[index_j]) 
+            * particles_->porous_solid_.PairNumericalDamping(strain_rate,  smoothing_length_);
+    
+        //three parts for the momentum increment
+        total_momentum_increment += (Stress_[index_i] + Stress_[index_j])* gradw_ijV_j_
+                                    +  numerical_dissipation_factor_  * numerical_stress_ij * weight * gradw_ijV_j_ ;
+    
+        total_momentum_increment -= (outer_fluid_velocity_relative_fluid_flux_[index_i] + outer_fluid_velocity_relative_fluid_flux_[index_j])
+                                     *gradw_ijV_j_;
+    }
```
Comment (DongWuTUM) [inline tests/user_examples/extra_src/shared/porous_media_dynamics.cpp:92]: If is the following better?
total_momentum_increment += (Stress_[index_i] + Stress_[index_j] 
\+  numerical_dissipation_factor_  * numerical_stress_ij * weight 
\- outer_fluid_velocity_relative_fluid_flux_[index_i] - outer_fluid_velocity_relative_fluid_flux_[index_j]) 
\* gradw_ijV_j_;

File: tests/user_examples/extra_src/shared/porous_media_dynamics.cpp (hunk 3 above)
Comment (Xiaojingtang1234) [reply tests/user_examples/extra_src/shared/porous_media_dynamics.cpp:95]: resolved.

**Review by DongWuTUM — APPROVED** [review]

**Pull request merged**